Lib/importlib/_bootstrap_external.py

```diff
@@ -208,12 +208,8 @@ def _write_atomic(path, data, mode=0o666):
     try:
         # We first write data to a temporary file, and then use os.replace() to
         # perform an atomic rename.
-        with _io.FileIO(fd, 'wb') as file:
-            bytes_written = file.write(data)
-        if bytes_written != len(data):
-            # Raise an OSError so the 'except' below cleans up the partially
-            # written file.
-            raise OSError("os.write() didn't write the full pyc file")
+        with _io.FileIO(fd, 'wb') as file, _io.BufferedWriter(file) as writer:
```

Comment:
I think just `with open(fd, 'wb') as file` should work here. 

Could technically pass the `_os.open` as an `opener` (`open(opener=_os.open(...))`) but I don't think that actually makes the code simpler/more straightforward

Reply:
Sure, done that. It did mean killing the tests, but I don't think we need them at this point.

Reply:
https://github.com/cmaloney/cpython/commit/8bf5cedc81d68a1b7520f8f95a04980488a5b8c0 should do it (builtins aren't available in this piece of code; the `open` builtin maps to `_io.open` generally)

I think it's valuable to add/keep the atomic test to make sure future optimizers keep the guarantees this is adding.

Unfortunately the Python code doesn't call `_os.write` so harder to mock but I think you should be able to use [`unittest.mock.patch`](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.patch) to do something like:

```python3
with patch.object(io.FileIO, 'write', write):
    ...
```

 (Will probably take a little iteration to get right)

Reply:
Ah, right, yeah. That's simple. Done.

Reply:
> Unfortunately the _io implementation doesn't call _os.write

If we keep the same patch approach of replacing `_io` with `_pyio`, the same tests work.

Thanks for the `_io.open` hint.